<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html>
  <head>
    <meta http-equiv="content-type" content="text/html; charset=ISO-8859-1">
    <title>CS481</title>
  </head>
  <body>
    <h1>CS 481: Bioinformatics Algorithms</h1>
    <p>...&nbsp; more like:</p>
    <p><sub>(bio)</sub>informatics <strong>ALGORITHMS</strong></p>
    <h3>Spring 2025<br>
    </h3>
    <ul>
      <li>Instructor: <a href="http://www.cs.bilkent.edu.tr/%7Ecalkan/">Can
          Alkan</a></li>
      <li>TA: </li>
      <ul>
        <li><a href="https://www.alkanlab.org/team/akmuhammet-ashyralyyev/">Akmuhammet
            Ashyralyyev</a></li>
        <li><a href="https://www.alkanlab.org/team/z%C3%BClal-bing%C3%B6l/">Zülal
            Bingöl</a></li>
      </ul>
      <li>Grader: </li>
      <ul>
        <li><a href="https://www.alkanlab.org/team/ecem-ilg%C3%BCn/">Ecem &#304;lgün</a></li>
      </ul>
      <li>Class hours:
        <ul>
          <li>Tue 13:30-15:20 - Fri 09:30-10:20 </li>
        </ul>
      </li>
      <li>Class room: B-Z08</li>
      <li>Office hour:&nbsp; by appointment. </li>
      <ul>
        <li>Public calendar (changes frequently): <a href="http://cs.bilkent.edu.tr/%7Ecalkan/calendar.html">http://cs.bilkent.edu.tr/~calkan/calendar.html</a></li>
      </ul>
    </ul>
    <p> </p>
    <p><br>
    </p>
    <hr style="width: 100%; height: 2px;">NOTE:&nbsp; Some biology, molecular
    biology, genetics background would help, but <span style="font-weight: bold;">not</span>
    required. Basics regarding the topic will be covered in class.<br>
    <hr style="width: 100%; height: 2px;"><br>
    <strong>Prerequisites:</strong> Elementary discrete mathematics, basic
    algorithms and data structures, and programming proficiency with, e.g.,
    C/C++/Java will be expected. Knowledge of elementary combinatorics and
    probability, fundamental algorithms for sorting, searching, hashing and
    dynamic programming, elementary graph algorithms would be very helpful.<br>
    <br>
    <strong>Textbook:</strong> none required.<br>
    <ul>
      <li>Recommended Material</li>
      <ul>
        <li>Genome-Scale Algorithm Design, Veli Mäkinen, Djamal Belazzougui,
          Fabio Cunial, Alexandru I. Tomescu, 2015, Cambridge University Press</li>
        <ul>
          <li><a href="https://www.cambridge.org/core/books/genomescale-algorithm-design/F51FF2C13B97ADB7C04B3D4345A3A385">Available
              in full through Bilkent library</a></li>
        </ul>
        <li>An Introduction to Bioinformatics Algorithms (Computational
          Molecular Biology), Neil Jones and Pavel Pevzner, MIT Press, 2004</li>
        <li>Bioinformatics algorithms : an active learning approach, Phillip
          Compeau and Pavel Pevzner, Active Learning Publishers; 2018</li>
        <ul>
          <li><a href="https://www.bioinformaticsalgorithms.org/">Some chapters
              available online</a></li>
        </ul>
      </ul>
      <ul>
        <li>Biological Sequence Analysis: Probabilistic Models of Proteins and
          Nucleic Acids, Richard Durbin, Sean R. Eddy, Anders Krogh, Graeme
          Mitchison, Cambridge University Press</li>
        <ul>
          <li><a href="https://www.cambridge.org/core/books/biological-sequence-analysis/921BB7B78B745198829EF96BC7E0F29D">Available
              in full through Bilkent library</a></li>
        </ul>
      </ul>
      <ul>
        <li>Bioinformatics: The Machine Learning Approach, Second Edition,
          Pierre Baldi, Soren Brunak, MIT Press</li>
      </ul>
      <ul>
        <li>Algorithms on Strings, Trees, and Sequences: Computer Science and
          Computational Biology, Dan Gusfield, Cambridge University Press</li>
        <ul>
          <li><a href="https://www.cambridge.org/core/books/algorithms-on-strings-trees-and-sequences/F0B095049C7E6EF5356F0A26686C20D3">Available
              in full through Bilkent library</a></li>
        </ul>
      </ul>
    </ul>
    <p><strong>Grading:</strong> There will be one midterm exam (25%) and a
      final exam (35%). A further 30% of the final grade will be based on
      homework/programming assignments, and 10% will be based on quizzes.&nbsp;<strong></strong></p>
    <p><strong><span style="color: red;"> Grading Policy: Note that we do not
          discuss with students about grades. Therefore, we will not answer any
          questions about the passing grades and/or students' requests for
          passing the course. Any emails sent to this effect will be omitted.</span></strong></p>
    <hr style="width: 100%; height: 2px; color: black;"> <strong>Planned
      Lecture Cancellations</strong><br>
    The following lectures are most likely to be canceled due to conference
    travels. <br>
    <ul>
      <li>April 25, 2025</li>
      <li>April 29, 2025</li>
    </ul>
    <hr style="color: black;">
    <h3>Course material (updated as of Fall 2022 - subject to further updates)</h3>
    <p> Note: Some slides are adapted from <a href="https://www.bioinformaticsalgorithms.org/">https://www.bioinformaticsalgorithms.org/</a>
      <em><br>
      </em></p>
    <p><em>All pseudocodes in the slides are provided as guidelines, not meant
        for direct "copy-paste" implementation.</em> </p>
    <ul>
      <li><a href="https://docs.google.com/presentation/d/19aCxS98IuZg-78yT6bAzXAtvQT-1Z6Cgp3b9o-7OYwc/edit?usp=sharing">0-
          Introduction</a></li>
      <li><a href="https://docs.google.com/presentation/d/1rU0DhBg6yVXbfEtNuK73pjc1yWSPG9OYnGNH-b-kklQ/edit?usp=sharing"><span
            style="color: #0000ee;">0b- Short note on homework implementations</span></a></li>
      <li><a href="https://docs.google.com/presentation/d/1VAtHqNHatRwnoalnz_T1AGV5oX3H6LwJ6haS-qh1bO4/edit?usp=sharing">1-
          Motif Finding &amp; Algorithm Design Techniques</a></li>
      <li><a href="https://docs.google.com/presentation/d/1H1DplgCxIvn6zsBW4CN3A83Mop66jFsXZHZQUuJnxmc/edit?usp=sharing">2-
          Exact String Matching - Knuth-Morris-Pratt and Boyer-Moore</a></li>
      <li><a href="https://docs.google.com/presentation/d/1aiH-mRVMo2Cq5Wppe1VhxbFn3sD8X9yjcUs9NHytEJM/edit?usp=sharing">3-
          Exact String Matching - Rabin-Karp, Finite Automata. Shift/And<br>
        </a></li>
      <li><a href="https://docs.google.com/presentation/d/1HF5j9uaX6VLE7viBwYPVnuli1X-qGGuaoTePs49rH5k/edit?usp=sharing">4-
          Pattern Matching - Hashing, keyword trees, and suffix trees, suffix
          arrays, Aho-Corasick<br>
        </a></li>
      <li><a href="https://docs.google.com/presentation/d/1sCPNQwVAD9A-Dx5U8akcJh8lesxheuF-mJ4wYoM_XpU/edit?usp=sharing">5-
          Burrows-Wheeler Transformation and Ferragina-Manzini index</a></li>
      <li><a href="https://docs.google.com/presentation/d/1quxF2pbCO1b_jRNzY5qHajqLADv_JW2y7tZBWqo2K4U/edit?usp=sharing">6-
          Dynamic programming, Manhattan grid, edit distance, longest common
          subsequence</a></li>
      <li><a href="https://docs.google.com/presentation/d/15urWu7TXmdQ5ocRs-k2O5zGPXaHu-oehhpGND3S_SbE/edit?usp=sharing">7-
          Global alignment, affine gap penalties, local alignment,
          Needleman-Wunsch and Smith-Waterman</a></li>
      <li style="margin-left: 40px;"><a href="https://docs.google.com/spreadsheets/d/1-V6n0QG81SlYpYHx4dVGzjQ4DO90hnJRsKit_jQsr24/edit?usp=sharing"><span
            style="color: #0000ee;">Needleman-Wunsch with affine gaps sample</span></a></li>
      <li style="margin-left: 40px;">Smith-Waterman samples</li>
      <li style="margin-left: 80px;"><a href="pdfslides/Smith-Waterman%20Algorithm2_edited-static.xls"><span
            style="  color: #0000ee;">Static</span></a></li>
      <li style="margin-left: 80px;"><span style="   color: #0000ee;"><a href="pdfslides/Smith-Waterman%20Algorithm2_edited.xls">Random
            sequence generator</a><br>
        </span></li>
      <li><a href="https://docs.google.com/presentation/d/1gbBN53wEUF1s_0bhZsALrBhATP0XTWXfe8a-IOpgJkY/edit?usp=sharing">8-
          Banded alignment, linear space alignment, block edit distance, Four
          Russians method</a></li>
      <li><a href="https://docs.google.com/presentation/d/1MXi5hZvysIpFuyTlEYR9MHVCVzIB5jJ3WmnjNcaopmI/edit?usp=sharing">9-
          Multiple sequence alignment</a></li>
      <li><a href="https://docs.google.com/presentation/d/1c626-ISa8nqAHmxN0_Pf7r6LUAC9KyTkYrfAX_a3xH8/edit?usp=sharing">10-
          Phylogenetic tree construction: Guide trees, evolutionary trees -
          Neighbor-Joining and UPGMA, parsimony, Sankoff and Fitch algorithms</a></li>
      <li><a href="https://docs.google.com/presentation/d/15yK7qbAEayvCV89eCgdUWdgyHLr6gZZ8uZCM9cL2G9w/edit?usp=sharing">11
          - Heuristic sequence search. Short introduction to BLAST. Hash table
          indexes, minimizers. Chaining</a></li>
      <li><span style=" color: #0000ee;"><span style="color: black;">12- <a href="https://docs.google.com/presentation/d/15yK7qbAEayvCV89eCgdUWdgyHLr6gZZ8uZCM9cL2G9w/edit?usp=sharing">MEMs,
              MUMs, BWA-MEM, minimap2</a>,</span> </span><a href="https://docs.google.com/presentation/d/1SoscAOz1Jra-FkesX6ai-u7uWm4wOXolUqWiqi-b1P0/edit?usp=sharing">K-mer
          data structures</a></li>
      <!--
      <li>13- Alignment-free k-mer composition analysis. Minimum        perfect hashing, MinHash, Jaccard Index</li>-->
      <li><a href="https://docs.google.com/presentation/d/1SoscAOz1Jra-FkesX6ai-u7uWm4wOXolUqWiqi-b1P0/edit?usp=sharing">14-
          Graphs in genome analysis. OLC, de Bruijn, string graphs.</a> Aligning
        reads to graphs</li>
      <li><a href="https://docs.google.com/presentation/d/1mRadHSAGcAyCR_ZE1iXeodn84rJg1JImsi2cXdSqI58/edit?usp=sharing">15-
          Applications: short introduction to genome sequencing. Current
          platforms and data types. Standard file formats</a></li>
      <li>16- Applications: programming libraries, application-specific
        programming languages</li>
      <p><span style="color: #0000ee;"></span><a href="https://docs.google.com/presentation/d/1kpaKQd6_3H9RiI5hPLiBqYStAl9BARuCpCkTdx2wXIM/edit?usp=sharing"><span
            style="color: #0000ee;"></span></a><a href="https://docs.google.com/presentation/d/107lChq6QYaXyRKLZbX05PceRAFceUGcfyDZCno4oQqQ/edit?usp=sharing"><span
            style=" color: #0000ee;"></span></a></p>
    </ul>
    <!--
    <li><span style="  color: #0000ee;"><br>      </span></li>    <li><a href="https://docs.google.com/presentation/d/1mRadHSAGcAyCR_ZE1iXeodn84rJg1JImsi2cXdSqI58/edit?usp=sharing"><span          style="  color: #0000ee;">17- Introduction to DNA sequencing</span></a></li>    <li><a href="https://docs.google.com/presentation/d/1RQT5tom3fN7wNLreOVqJ5Hkzp7jboQ2H7wBc59CIGaU/edit?usp=sharing"><span          style="   color: #0000ee;">18- High throughput sequencing,          compression, read mapping</span></a></li>    <li><span style="    color: #0000ee;"><a href="pdfslides/20-cs481-intro_to_assembly.pdf">19-          Introduction to genome assembly</a><br>      </span></li>-->
    <hr style="color: black;"><strong><span style="text-decoration: underline;"></span></strong>
    <p><br>
    </p>
    <h3>Key dates (tentative, subject to change)</h3>
    <ul>
      <li>Midterm:&nbsp; March 25, 2025 -- during class hours</li>
      <li>Quizzes:</li>
      <ul>
        <li>Quiz 1: February 11, 2025</li>
        <li>Quiz 2: March 4, 2025</li>
        <li>Quiz 3:March 18, 2025</li>
        <li>Quiz 4: April 15, 2025</li>
        <li>Quiz 5: May 13, 2025</li>
      </ul>
      <li>Homeworks:</li>
      <ul>
        <li>HW0:</li>
        <ul>
          <li>Assigned: </li>
          <li>Deadline: </li>
        </ul>
      </ul>
      <ul>
        <li>HW1:</li>
        <ul>
          <ul>
          </ul>
          <li>Assigned: </li>
          <ul>
          </ul>
          <li>Deadline: </li>
        </ul>
        <li>HW2:</li>
        <ul>
          <li>Assigned: </li>
          <li>Deadline: </li>
        </ul>
        <li>HW3:</li>
        <ul>
          <li>Assigned: </li>
          <li>Deadline: </li>
        </ul>
        <li>HW4:</li>
        <ul>
          <li>Assigned: </li>
          <li>Deadline: </li>
        </ul>
        <li>HW5:</li>
        <ul>
          <li>Assigned: </li>
          <li>Deadline: </li>
        </ul>
      </ul>
    </ul>
    <h3>Important information for exams</h3>
    <ul>
      <li>You will be given the opportunity to bring cheat sheet to the midterm
        and final exams (not the quizzes). DO NOT try to memorize anything,
        instead grasp the logic behind each algorithm &amp; data structure.</li>
      <li>Write legibly, and make your answers clear.</li>
      <li>Unintelligible answers may fail to receive credit.</li>
      <li>In the case of a "design an algorithm" question, you are expected to
        give only pseudocode. Any "real" code may not be graded.</li>
    </ul>
    <p></p>
    <h3>Frequently Asked Questions</h3>
    <p></p>
    <ul>
      <li>I need at least grade (C/C+/B etc ..) to be able to graduate or keep
        my status at a certain level (satisfactory, honor, high honor, etc.).
        Can you please increase my grade?</li>
      <ul>
        <li>No.</li>
      </ul>
      <li>I am likely receiving an F and fail the course. Can you lower the F
        threshold?</li>
      <ul>
        <li>No.</li>
      </ul>
      <li>I need at least some specific grade to keep my cGPA above some
        threshold. Can you please give me at least (something)?</li>
      <ul>
        <li>See the question above.</li>
      </ul>
      <li>I think I deserve more partial credits from question X in exam.</li>
      <ul>
        <li>No.</li>
      </ul>
      <li>Which topics should I prioritize to study more for the midterm and/or
        final exam?</li>
      <ul>
        <li>All of them.<br>
        </li>
      </ul>
    </ul>
    <hr style="width: 100%; height: 2px;"><br>
  </body>
</html>
